daemon/src/fan_out.rs

```diff
@@ -1,32 +1,92 @@
+use anyhow::Result;
 use xtra::prelude::MessageChannel;
 
 /// A fan-out actor takes every incoming message and forwards it to a set of other actors.
-pub struct Actor<M: xtra::Message<Result = ()>> {
+pub struct Actor<M: xtra::Message<Result = Result<()>>> {
     receivers: Vec<Box<dyn MessageChannel<M>>>,
 }
 
-impl<M: xtra::Message<Result = ()>> Actor<M> {
+impl<M: xtra::Message<Result = Result<()>>> Actor<M> {
     pub fn new(receivers: &[&dyn MessageChannel<M>]) -> Self {
         Self {
             receivers: receivers.iter().map(|c| c.clone_channel()).collect(),
         }
     }
 }
 
-impl<M> xtra::Actor for Actor<M> where M: xtra::Message<Result = ()> {}
+impl<M> xtra::Actor for Actor<M> where M: xtra::Message<Result = Result<()>> {}
 
 #[async_trait::async_trait]
 impl<M> xtra::Handler<M> for Actor<M>
 where
-    M: xtra::Message<Result = ()> + Clone + Sync + 'static,
+    M: xtra::Message<Result = Result<()>> + Clone + Sync + 'static,
 {
-    async fn handle(&mut self, message: M, _: &mut xtra::Context<Self>) {
+    async fn handle(&mut self, message: M, _: &mut xtra::Context<Self>) -> Result<()> {
+        let mut errors = Errors::new();
         for receiver in &self.receivers {
-            if receiver.send(message.clone()).await.is_err() {
-                // Should ideally remove from list but that is unnecessarily hard with Rust
-                // iterators
-                tracing::error!("Actor disconnected, cannot send message");
+            match receiver.send(message.clone()).await {
+                Ok(Err(e)) => errors.add(e),
+                Err(e) => errors.add(e),
+                Ok(_) => continue,
             }
         }
+
+        if !errors.0.is_empty() {
+            return Err(errors.into());
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub struct Errors(Vec<anyhow::Error>);
+
+impl Errors {
+    fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    fn add<E: Into<anyhow::Error>>(&mut self, e: E) {
+        self.0.push(anyhow::anyhow!(e))
+    }
+}
+
+impl Default for Errors {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl std::fmt::Display for Errors {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let last = match self.0.last() {
+            Some(e) => e,
+            None => return Ok(()),
+        };
+
+        for error in self.0[..(self.0.len() - 1)].iter() {
+            (write!(f, "{} + ", error))?
+        }
+
+        (write!(f, "{}", last))?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn errors_display() -> Result<()> {
```

Comment:
Can we assert the actual display impl here?

Reply:
Sorry, I was just testing things for myself. Do you think it's worth it to keep this test?

Reply:
Well, I am not convinced we need this entire construct? Do we need to change `fan_out` to return a result? Note that I think we can also just drop the usage of `fan_out` if it is too painful.